pkg/render/common/components/components.go

```diff
@@ -0,0 +1,159 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package components
+
+import (
+	"fmt"
+
+	"github.com/tigera/operator/pkg/common"
+	"github.com/tigera/operator/pkg/components"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+var log = logf.Log.WithName("components")
+
+// replicatedPodResource contains the overridable data for a Deployment or DaemonSet.
+type replicatedPodResource struct {
+	labels          map[string]string
+	annotations     map[string]string
+	minReadySeconds *int32
+	podTemplateSpec *corev1.PodTemplateSpec
+}
+
+// applyReplicatedPodResourceOverrides takes the given replicated pod resource data and applies the overrides.
+func applyReplicatedPodResourceOverrides(r *replicatedPodResource, overrides components.ReplicatedPodResourceOverrides) *replicatedPodResource {
+	if metadata := overrides.GetMetadata(); metadata != nil {
+		if len(metadata.Labels) > 0 {
+			if r.labels == nil {
+				r.labels = make(map[string]string)
+			}
+			common.MergeMaps(metadata.Labels, r.labels)
+		}
+		if len(metadata.Annotations) > 0 {
+			if r.annotations == nil {
+				r.annotations = make(map[string]string)
+			}
+			common.MergeMaps(metadata.Annotations, r.annotations)
+		}
+	}
+	if minReadySeconds := overrides.GetMinReadySeconds(); minReadySeconds != nil {
+		r.minReadySeconds = minReadySeconds
+	}
+	if podTemplateMetadata := overrides.GetPodTemplateMetadata(); podTemplateMetadata != nil {
+		if len(podTemplateMetadata.Labels) > 0 {
+			if r.podTemplateSpec.Labels == nil {
+				r.podTemplateSpec.Labels = make(map[string]string)
+			}
+			common.MergeMaps(podTemplateMetadata.Labels, r.podTemplateSpec.Labels)
+		}
+		if len(podTemplateMetadata.Annotations) > 0 {
+			if r.podTemplateSpec.Annotations == nil {
+				r.podTemplateSpec.Annotations = make(map[string]string)
+			}
+			common.MergeMaps(podTemplateMetadata.Annotations, r.podTemplateSpec.Annotations)
+		}
+	}
+	if initContainers := overrides.GetInitContainers(); initContainers != nil {
+		mergeContainers(r.podTemplateSpec.Spec.InitContainers, initContainers)
+	}
+	if containers := overrides.GetContainers(); containers != nil {
+		mergeContainers(r.podTemplateSpec.Spec.Containers, containers)
+	}
+	if affinity := overrides.GetAffinity(); affinity != nil {
+		r.podTemplateSpec.Spec.Affinity = affinity
+	}
+	if nodeSelector := overrides.GetNodeSelector(); nodeSelector != nil {
+		r.podTemplateSpec.Spec.NodeSelector = nodeSelector
+	}
+	if tolerations := overrides.GetTolerations(); tolerations != nil {
+		r.podTemplateSpec.Spec.Tolerations = tolerations
+	}
+
+	return r
+}
+
+// ApplyDaemonSetOverrides applies the overrides to the given DaemonSet.
+// Note: overrides must not be nil pointer.
+func ApplyDaemonSetOverrides(ds *appsv1.DaemonSet, overrides components.ReplicatedPodResourceOverrides) *appsv1.DaemonSet {
```

Comment:
This function takes a pointer and returns a pointer to the same object - might be simpler to just always rely on the side effect? Or to make a copy at the top so we're not modifying the original object.

Reply:
Looks like you don't actually use the return value, so simply removing the return would do it.

Reply:
Fixed cf665ca31fcb7ccfd197bfee40136a07b90fd6b2